c8y_configuration_plugin --init and c8y_log_plugin --init are overwriting exiting files

PradeepKiruvale opened the pull request:
Signed-off-by: Pradeep Kumar K J <pradeepkumar.kj@softwareag.com>

## Proposed changes

`tedge/tedge_mapper/tedge_agent/c8y_config_plugin/c8y_log_plugin --init` must create the config files only if the config files does not exists.

Changed the signature of the `create_file_with_user_group` function to take `default_content` and write it to the file.

## Types of changes

<!--
What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_
-->

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

https://github.com/thin-edge/thin-edge.io/issues/1189

## Checklist

<!--
_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to ask.
We're here to help! This is simply a reminder of what we are going to look for
before merging your code._
-->

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

initard reviewed (commented):
Looks good to me. Just a two things: 

1. The name of the enum could be changed to something else so that when we create a new directory there is no return Ok(FileCreateStatus::CreateNew)? 
2. Is it possible to have a test to check existing files are not overwritten?

PradeepKiruvale commented on crates/common/tedge_utils/src/file.rs:
```diff
@@ -97,27 +100,35 @@ impl PermissionEntry {
+                Ok(FileCreateStatus::CreateNew)
+            }
+            Err(e) if e.kind() == io::ErrorKind::AlreadyExists => {
+                Ok(FileCreateStatus::AlreadyExists)
```
Returning the status is not necessary, so removed this code.

PradeepKiruvale commented on crates/common/tedge_utils/src/file.rs:
```diff
@@ -1,11 +1,16 @@
 use users::{get_group_by_name, get_user_by_name};
 
+#[derive(Debug)]
+pub enum FileCreateStatus {
+    CreateNew,
+    AlreadyExists,
+}
+
```
File create status has to be returned only for the `file`, not for the `directory`. So, I removed this code.

didier-wenzek replied:
`CreationStatus` could be a better name.  `CreatedNew` would also be better than `CreateNew` that sounds as a command rather than a status.

PradeepKiruvale replied:
I completely get rid of this.

PradeepKiruvale commented:
> Looks good to me. Just a two things:
> 
> 1. The name of the enum could be changed to something else so that when we create a new directory there is no return Ok(FileCreateStatus::CreateNew)?
The file create status is required only for the `file`, not for the `directory`. So, I removed the code.
> 2. Is it possible to have a test to check existing files are not overwritten?
Yes, added a test to check if the file already exists.

didier-wenzek commented on crates/common/tedge_utils/src/file.rs:
```diff
@@ -37,13 +42,13 @@ pub fn create_directory_with_user_group(
     mode: u32,
 ) -> Result<(), FileError> {
     let perm_entry = PermissionEntry::new(Some(user.into()), Some(group.into()), Some(mode));
-    let () = perm_entry.create_directory(Path::new(dir))?;
+    let _ = perm_entry.create_directory(Path::new(dir))?;
     Ok(())
```
See https://github.com/thin-edge/thin-edge.io/pull/1208 for a more idiomatic style.

didier-wenzek commented on crates/common/tedge_utils/src/file.rs:
```diff
@@ -111,13 +114,19 @@ impl PermissionEntry {
 
-    fn create_file(&self, file: &Path) -> Result<(), FileError> {
-        match File::create(file) {
+    fn create_file(&self, file: &Path) -> Result<FileCreateStatus, FileError> {
```
This function deserves a doc comment.

PradeepKiruvale replied:
Updated the doc comment.

didier-wenzek commented on plugins/c8y_log_plugin/src/main.rs:
```diff
@@ -244,17 +246,20 @@ fn create_init_logs_directories_and_files(
-        &format!("{config_dir}/{DEFAULT_PLUGIN_CONFIG_FILE}"),
-        logs_dir,
-    )?;
+    if let FileCreateStatus::CreateNew = status {
+        // append default content to c8y-log-plugin.toml
+        create_default_log_plugin_file(
+            &format!("{config_dir}/{DEFAULT_PLUGIN_CONFIG_FILE}"),
+            logs_dir,
+        )?;
+    }
```
I'm not really happy with this API. It's okay but a bit complex for little added value: one create an empty file, then close it, and if actually new reopen it to add some content if the file were actually new. Why not a create function with a default content to be used if the file is new.

rina23q reviewed (commented):
There is a bug.

Precondition:
* `/etc/tedge` directory exist.
* `/etc/tedge/c8y` directory doesn't exist.

Run:
```
sudo ./c8y_log_plugin --init
```

Result:
```
sudo ./c8y_log_plugin --init
2022-06-23T12:17:21.383339143Z  INFO c8y_log_plugin: Creating supported operation files
Error: Unable to open file: /etc/tedge/c8y/c8y-log-plugin.toml. Error: Invalid argument (os error 22)
```

Note: this doesn't happen for c8y_configuration_plugin.

PradeepKiruvale commented:
> There is a bug.
> 
> Precondition:
> 
> * `/etc/tedge` directory exist.
> * `/etc/tedge/c8y` directory doesn't exist.
> 
> Run:
> 
> ```
> sudo ./c8y_log_plugin --init
> ```
> 
> Result:
> 
> ```
> sudo ./c8y_log_plugin --init
> 2022-06-23T12:17:21.383339143Z  INFO c8y_log_plugin: Creating supported operation files
> Error: Unable to open file: /etc/tedge/c8y/c8y-log-plugin.toml. Error: Invalid argument (os error 22)
> ```
You should run as `tedge` user.
`sudo runuser -u tedge -- ./target/debug/c8y_log_plugin --init`
> 
> Note: this doesn't happen for c8y_configuration_plugin.

Its not happening here because the `c8y` directory is getting created by `root`, as eblow
`create_directory_with_user_group(&format!("{config_dir}/c8y"), "root", "root", 0o775)?;`
So, we must make some decision here  about which owner the `root` or `tedge` user owns this folder.

didier-wenzek commented:
> > You should run as tedge user.
sudo runuser -u tedge -- ./target/debug/c8y_log_plugin --init

> I know it's not related to this PR's scope, but I'm totally surprised the design is like this. I see no reason `c8y_log_plugin` must be run as `tedge` user. 

I fully agree with @rina23q. In practice `c8y_log_plugin` must run as `root` to have access to restricted logs like those of mosquitto.

>> Its not happening here because the c8y directory is getting created by root, as eblow
create_directory_with_user_group(&format!("{config_dir}/c8y"), "root", "root", 0o775)?;
So, we must make some decision here about which owner the root or tedge user owns this folder.

> I also checked the code. On one hand, `c8y_configuration_plugin` creates `/etc/tedge/c8y` owned by `root`, and `c8y_log_plugin` creates `/etc/tedge/c8y` owned by `tedge`.
> In my opinion, I support to make all owned by `root`. Our systemd file for the log plugin starts the process as `root`, I see no reason why we need to make the plugin file and c8y directory owned by `tedge`. 
https://github.com/thin-edge/thin-edge.io/blob/main/configuration/init/systemd/c8y-log-plugin.service

Here, I disagree with @rina23q. I would prefer to have these files and directories owned by `tedge`. Otherwise this will add another constraint to an operation plugins that declares supported operation.

PradeepKiruvale commented:
> > > You should run as tedge user.
> > > sudo runuser -u tedge -- ./target/debug/c8y_log_plugin --init
> 
> > I know it's not related to this PR's scope, but I'm totally surprised the design is like this. I see no reason `c8y_log_plugin` must be run as `tedge` user.
> 
> I fully agree with @rina23q. In practice `c8y_log_plugin` must run as `root` to have access to restricted logs like those of mosquitto.
> 
> > > Its not happening here because the c8y directory is getting created by root, as eblow
> > > create_directory_with_user_group(&format!("{config_dir}/c8y"), "root", "root", 0o775)?;
> > > So, we must make some decision here about which owner the root or tedge user owns this folder.
> 
> > I also checked the code. On one hand, `c8y_configuration_plugin` creates `/etc/tedge/c8y` owned by `root`, and `c8y_log_plugin` creates `/etc/tedge/c8y` owned by `tedge`.
> > In my opinion, I support to make all owned by `root`. Our systemd file for the log plugin starts the process as `root`, I see no reason why we need to make the plugin file and c8y directory owned by `tedge`.
> > https://github.com/thin-edge/thin-edge.io/blob/main/configuration/init/systemd/c8y-log-plugin.service
> 
> Here, I disagree with @rina23q. I would prefer to have these files and directories owned by `tedge`. Otherwise this will add another constraint to an operation plugins that declares supported operation.

I agree with @didier-wenzek. Even in the `c8y-configuration-plugin`, some files (c8y_UploadConfigFile, c8y_DownloadConfigFile, operations/c8y) are owned by `tedge`. So, let's get rid of this inconsistency by changing the owner for all the files to `tedge`. I will make the changes accordingly

didier-wenzek commented:
> Now we are talking about `/etc/tedge/c8y` directory

I missed that point.

> > Otherwise this will add another constraint to an operation plugins that declares supported operation.

> I don't get this point fully.

Even if an operation plugin doesn't need to run as root, it will have to create its configuration file as root.

But I see a real benefit of not having the `--init` option creating this directory as owned by `tedge`: it will be really good to move away from that model where thin-edge works only with a user named `tedge`. 

A solution can be to have the directory owned by `root` but allowing anybody to add a configuration file there:

```
$ ls -ld /etc/tedge/c8y
drwxrwxrwt 8 root root 4096 Jun 24 10:32 /etc/tedge/c8y
```

rina23q commented on plugins/c8y_configuration_plugin/src/main.rs:
```diff
@@ -204,25 +203,23 @@ fn init(cfg_dir: PathBuf) -> Result<(), anyhow::Error> {
 
 fn create_operation_files(config_dir: &str) -> Result<(), anyhow::Error> {
     create_directory_with_user_group(&format!("{config_dir}/c8y"), "root", "root", 0o775)?;
```
According to the Didier's comment https://github.com/thin-edge/thin-edge.io/pull/1217#issuecomment-1165430950, the `/etc/tedge/c8y` directory permission also should be changed.

```suggestion
    create_directory_with_user_group(&format!("{config_dir}/c8y"), "root", "root", 0o1777)?;
```

PradeepKiruvale replied:
done

rina23q commented on plugins/c8y_log_plugin/src/main.rs:
```diff
@@ -186,27 +182,6 @@ fn init(config_dir: &Path, logs_dir: &Path) -> Result<(), anyhow::Error> {
-
 /// for the log plugin to work the following directories and files are needed:
 ///
 /// Directories:
```
It's not related to your PR, but the comments below contains inaccurate.

Wrong:
```
/// - CONFIG_DIR/c8y/log/c8y-log-plugin.toml
```

Correct:
```
/// - CONFIG_DIR/c8y/c8y-log-plugin.toml
```

PradeepKiruvale replied:
done

rina23q commented on plugins/c8y_log_plugin/src/main.rs:
```diff
@@ -238,23 +213,26 @@ fn create_init_logs_directories_and_files(
         "tedge",
         "tedge",
         0o755,
```
Again not related to your PR, but makes no sense the supported operation file is executable. Also, it will keep consistency with `c8y-configuration-plugin`.

```suggestion
        0o644,
```

rina23q replied:
@PradeepKiruvale @initard It needs to be executable also for inotify? We need to decide either `0o644` or `0o755`.
It's correct that we are mixing up, on one hand, the config plugin creates supported operation files with `0o644`, on the other hand, the log plugin creates the supported operation file with `0o755`.

PradeepKiruvale replied:
@rina23q I spoke to @initard regarding `0644` permission to `c8y-log-plugin.toml`, and he said it will work.

rina23q commented on plugins/c8y_log_plugin/src/main.rs:
```diff
@@ -238,23 +213,26 @@ fn create_init_logs_directories_and_files(
     // creating c8y directory
-    create_directory_with_user_group(&format!("{config_dir}/c8y"), "tedge", "tedge", 0o755)?;
-    // creating c8y-log-plugin.toml
+    create_directory_with_user_group(&format!("{config_dir}/c8y"), "root", "root", 0o755)?;
```
According to the Didier's comment https://github.com/thin-edge/thin-edge.io/pull/1217#issuecomment-1165430950.
```suggestion
    create_directory_with_user_group(&format!("{config_dir}/c8y"), "root", "root", 0o1777)?;
```

PradeepKiruvale replied:
fixed

rina23q reviewed (commented):
Good API, but permissions need to be fixed.

didier-wenzek commented on plugins/c8y_log_plugin/src/main.rs:
```diff
@@ -242,22 +218,28 @@ fn create_init_logs_directories_and_files(
-        "tedge",
+        "root",
+        "root",
         0o775,
```
`664` please

PradeepKiruvale replied:
fixed

PradeepKiruvale commented on crates/common/tedge_utils/src/file.rs:
```diff
@@ -7,7 +7,7 @@ use users::{get_group_by_name, get_user_by_name};
 #[derive(Debug, PartialEq)]
 pub enum FileCreateStatus {
-    CreateNew,
+    CreatedNew,
```
Sure. From now onwards I will have one commit per change.

PradeepKiruvale commented on crates/common/tedge_utils/src/file.rs:
```diff
@@ -37,13 +42,13 @@ pub fn create_directory_with_user_group(
     let perm_entry = PermissionEntry::new(Some(user.into()), Some(group.into()), Some(mode));
-    let () = perm_entry.create_directory(Path::new(dir))?;
+    let _ = perm_entry.create_directory(Path::new(dir))?;
     Ok(())
```
done

PradeepKiruvale commented on crates/common/tedge_utils/src/file.rs:
```diff
@@ -37,13 +42,13 @@ pub fn create_directory_with_user_group(
     let perm_entry = PermissionEntry::new(None, None, Some(mode));
-    let () = perm_entry.create_directory(Path::new(dir))?;
+    let _ = perm_entry.create_directory(Path::new(dir))?;
     Ok(())
```
done

rina23q reviewed (commented):
Almost! Just fix the one point!

didier-wenzek commented on plugins/c8y_configuration_plugin/src/main.rs:
```diff
@@ -203,43 +202,43 @@ fn init(cfg_dir: PathBuf) -> Result<(), anyhow::Error> {
         "tedge",
         "tedge",
-        0o775,
+        0o1777,
```
Rina is correct.

PradeepKiruvale replied:
fixed

didier-wenzek reviewed (commented):
I will approve once fixed the latest point raised by @rina23q

rina23q reviewed (approved):
:+1:
